src/Storages/System/StorageSystemRemoteDataPaths.cpp

```diff
@@ -34,16 +164,121 @@ StorageSystemRemoteDataPaths::StorageSystemRemoteDataPaths(const StorageID & tab
     QueryProcessingStage::Enum /*processed_stage*/,
-    const size_t /*max_block_size*/,
+    const size_t max_block_size,
     const size_t /*num_streams*/)
 {
     storage_snapshot->check(column_names);
+    auto header = storage_snapshot->metadata->getSampleBlockWithVirtuals(getVirtualsList());
+    auto read_step = std::make_unique<ReadFromSystemRemoteDataPaths>(
+        context->getDisksMap(),
+        column_names,
+        query_info,
+        storage_snapshot,
+        context,
+        header,
+        max_block_size);
+    query_plan.addStep(std::move(read_step));
+}
+
+void ReadFromSystemRemoteDataPaths::initializePipeline(QueryPipelineBuilder & pipeline, const BuildQueryPipelineSettings & /*settings*/)
+{
+    const auto & header = getOutputStream().header;
+    auto source = std::make_shared<SystemRemoteDataPathsSource>(std::move(disks), header, max_block_size, context);
+    source->setStorageLimits(storage_limits);
+    processors.emplace_back(source);
+    pipeline.init(Pipe(std::move(source)));
+}
+
+bool SystemRemoteDataPathsSource::nextDisk()
+{
+    while (current_disk < static_cast<ssize_t>(disks.size()))
+    {
+        paths_stack.clear();
+        ++current_disk;
+
+        if (current_disk >= static_cast<ssize_t>(disks.size()))
+            break;
+
+        auto & current = paths_stack.emplace_back();
+
+        /// Add dirs that we want to traverse. It's ok if some of them don't exist because traversal logic handles
+        /// cases when children of a directory get deleted while traversal is running.
+        current.names.push_back({"store", nullptr});
+        current.names.push_back({"data", nullptr});
+        if (context->getSettingsRef().traverse_shadow_remote_data_paths)
+            current.names.push_back({"shadow", skipPredicateForShadowDir});
+
+        /// Start and move to the first file
+        current.position = -1;
+        if (nextFile())
+            return true;
+    }
+    return false;
+}
+
+bool SystemRemoteDataPathsSource::nextFile()
+{
+    while (true)
+    {
+        while (!paths_stack.empty())
+        {
+            auto & current = paths_stack.back();
+            ++current.position;
+            /// Move to the next child in the current directory
+            if (current.position < static_cast<ssize_t>(current.names.size()))
+                break;
+            /// Move up to the parent directory if this was the last child
+            paths_stack.pop_back();
+        }
+
+        /// Done with the current disk?
+        if (paths_stack.empty())
+            return false;
+
+        try
+        {
+            const auto & disk = disks[current_disk].second;
+            /// Stop if current path is a file
+            if (disk->isFile(getCurrentPath()))
+                return true;
+
+            /// If current path is a directory list its contents and step into it
+            std::vector<std::string> children;
+            disk->listFiles(getCurrentPath(), children);
```

Comment:
Looks like now it also lists garbage dirs that are being removed: https://s3.amazonaws.com/clickhouse-test-reports/0/26b9c9932824d85ee1706d5cc5b436cab67b4e8e/stateless_tests__tsan__s3_storage__[4_5].html 

Logs: https://pastila.nl/?0031f4fa/7595cfd174c73a3efa25b3098b92ccd5#vCpvH1BfXdHdovIOzn35Mw==